ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -425,17 +425,127 @@ public func catch<T, E, F>(handler: E -> SignalProducer<T, F>)(producer: SignalP
 	}
 }
 
+/// Returns a signal which sends all the values from each signal emitted from
+/// `producer`, waiting until each inner signal completes before beginning to
+/// send the values from the next inner signal.
+///
+/// If any of the inner signals emit an error, the returned signal will emit
+/// that error.
+///
+/// The returned signal completes only when `producer` and all signals
+/// emitted from `producer` complete.
+public func concat<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> SignalProducer<T, E> {
+	return SignalProducer { observer, disposable in
+		let state = Atomic(ConcatState<T, E>())
+		
+		/// Subscribes to the given signal producer.
+		var subscribeToSignalProducer: (SignalProducer<T, E> -> Void)?
+		
+		/// Sends completed to the subscriber if all signals are finished.
+		let completeIfAllowed = { (concatState: ConcatState<T, E>) -> Void in
+			if concatState.selfCompleted && concatState.activeDisposables.count == 0 {
+				sendCompleted(observer)
+				
+				// A strong reference is held to `subscribeToSignalProducer` until
+				// completion, preventing it from deallocating early.
+				subscribeToSignalProducer = nil
+			}
+		}
+		
+		subscribeToSignalProducer = Z { recur, signalProducer in
+			let serialDisposable = SerialDisposable()
+			let serialDisposableCompositeHandle = disposable.addDisposable(serialDisposable)
+			state.modify { oldState in
+				var newState = oldState
+				newState.activeDisposables.append(serialDisposable)
+				return newState
+			}
+			
+			serialDisposable.innerDisposable = signalProducer.start(next: { value in
+				sendNext(observer, value)
+			}, error: { error in
+				sendError(observer, error)
+			}, completed: {
+				var nextSignalProducer: SignalProducer<T, E>?
+				
+				serialDisposableCompositeHandle.remove()
+				state.modify { oldState in
+					var newState = oldState
+					newState.activeDisposables.filter { $0 !== serialDisposable }
+					if newState.queuedSignalProducers.count == 0 {
+						completeIfAllowed(newState)
+					} else {
+						nextSignalProducer = newState.queuedSignalProducers[0]
+						newState.queuedSignalProducers.removeAtIndex(0)
+					}
+					return newState
+				}
+				
+				if let nextSignalProducer = nextSignalProducer {
+					recur(nextSignalProducer)
+				}
+			})
+		}
+		
+		producer.startWithSignal { signal, signalDisposable in
+			let outerSignalDisposable = signal.observe(next: { innerSignalProducer in
```

Comment:
I apologize for not making this clear in the documentation, but you can use `signalDisposable` instead of this returned disposable for all intents and purposes. It's even preferable, because then the `Signal` can be disposed _before_ observation begins, if needed.